From 0f7001a792255a33436b44669a434fc20ff0774c Mon Sep 17 00:00:00 2001
From: Christian Carstensen <cc@nanl.de>
Date: Tue, 18 Nov 2014 17:29:29 +0100
Subject: [PATCH 2/9] udhcp: 4o6 message types and options by RFC 7341

---
 networking/udhcp/dhcp4o6.h | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/networking/udhcp/dhcp4o6.h b/networking/udhcp/dhcp4o6.h
index e81b0a5..e6d6de4 100644
--- a/networking/udhcp/dhcp4o6.h
+++ b/networking/udhcp/dhcp4o6.h
@@ -20,11 +20,11 @@ PUSH_AND_SET_FUNCTION_VISIBILITY_TO_HIDDEN
 #define MODE4o6_KERNEL_MODE  8
 
 /* DHCPv4o6 message types */
-#define D6_MSG_DHCPV4_QUERY        32
-#define D6_MSG_DHCPV4_RESPONSE     33
+#define D6_MSG_DHCPV4_QUERY        20
+#define D6_MSG_DHCPV4_RESPONSE     21
 
 /* DHCPv4o6 option */
-#define D6_OPT_DHCPV4_MSG          0xfe
+#define D6_OPT_DHCPV4_MSG          87
 
 /* send/recv/listen modes */
 #define SOCKET_NONE   0
-- 
2.1.3

